v2/component/launcher.go

```diff
@@ -482,6 +495,78 @@ func localPathForURI(uri string) (string, error) {
 	return "", fmt.Errorf("found URI with unsupported storage scheme: %s", uri)
 }
 
+func (l *Launcher) generateCacheKey(
+	inputs *pipeline_spec.ExecutorInput_Inputs,
+	outputs *pipeline_spec.ExecutorInput_Outputs,
+	cmdAgrs []string) (*cacheKey, error) {
+
+	cacheKey := cacheKey{
+		inputArtifactNames:   make(map[string]artifactNameList),
+		inputParameters:      make(map[string]pipeline_spec.Value),
+		outputArtifactsSpec:  make(map[string]pipeline_spec.RuntimeArtifact),
+		outputParametersSpec: make(map[string]string),
+	}
+
+	for inputArtifactName, inputArtifactList := range inputs.GetArtifacts() {
+		artifactNameList := artifactNameList{artifactNames: make([]string, 0)}
+		for _, artifact := range inputArtifactList.Artifacts {
+			artifactNameList.artifactNames = append(artifactNameList.artifactNames, artifact.GetName())
+		}
+		cacheKey.inputArtifactNames[inputArtifactName] = artifactNameList
+	}
+
+	for inputParameterName, inputParameterValue := range inputs.GetParameters() {
+		cacheKey.inputParameters[inputParameterName] = pipeline_spec.Value{
+			Value: inputParameterValue.Value,
+		}
+	}
+
+	for outputArtifactName, outputArtifactList := range outputs.GetArtifacts() {
+		if len(outputArtifactList.Artifacts) == 0 {
+			continue
+		}
+		// TODO: Support multiple artifacts someday, probably through the v2 engine.
+		outputArtifact := outputArtifactList.Artifacts[0]
+		outputArtifactWithUriWiped := pipeline_spec.RuntimeArtifact{
+			Name:     outputArtifact.GetName(),
+			Type:     outputArtifact.GetType(),
+			Metadata: outputArtifact.GetMetadata(),
+		}
+		cacheKey.outputArtifactsSpec[outputArtifactName] = outputArtifactWithUriWiped
+	}
+
+	for outputParameterName, _ := range outputs.GetParameters() {
+		outputParameter, ok := l.runtimeInfo.OutputParameters[outputParameterName]
+		if !ok {
+			return nil, fmt.Errorf("unknown parameter %q found in ExecutorInput_Outputs", outputParameterName)
+		}
+
+		cacheKey.outputParametersSpec[outputParameterName] = outputParameter.Type
+	}
+
+	cacheKey.containerSpec = containerSpec{
+		image:    l.options.ContainerImage,
+		cmdArgs : cmdAgrs,
+	}
+
+	return &cacheKey, nil
+
+}
+
+func (l *Launcher) generateFingerPrint(cacheKey cacheKey) (string, error) {
+	b, err := json.Marshal(cacheKey)
```

Comment:
Can this guarantee unordered items have a stable order? e.g. different input parameter names

Reply:
The json package always orders keys when marshalling. Specifically:

1. Maps have their keys sorted lexicographically
2. Structs keys are marshalled in the order defined in the struct

FYI: https://stackoverflow.com/questions/18668652/how-to-produce-json-with-sorted-keys-in-go and https://golang.org/pkg/encoding/json/#Marshal.